File: 77-combinations/combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> combi;
        backtrack(n,k,res,1,combi);
        return res;

    }
    void backtrack(int n,int k,vector<vector<int>>& res,int cur,vector<int>& comb){
            if(comb.size()==k){
                res.push_back(comb);
                return ;
            }
            for(int i=cur;i<=n;i++){
                comb.push_back(i);
                backtrack(n,k,res,i+1,comb);
                comb.pop_back();
            }
    }
};

File: 77-combinations/combinations.java
class Solution {
    List<List<Integer>> ans=new ArrayList<>();
    public List<List<Integer>> combine(int n, int k) {
        backtrack(1,n,k,new ArrayList<>());
        return ans;
    }
    private void backtrack(int begin,int end,int k,List<Integer> arr){
        if(arr.size()==k){
            ans.add(new ArrayList<>(arr));
            return;
        }
        for(int i=begin;i<=end;i++){
            arr.add(i);
            backtrack(i+1,end,k,arr);
            arr.remove(arr.size()-1);
        }

    }
}